### ISSUE TITLE
[Bug][Param Manager] Decode lift in FuseDecodeTranspose may have some bugs.

### ISSUE THREAD

**Issue by LeiWang1999** [mlc-ai/mlc-llm#523]
Refer to this draft pr https://github.com/mlc-ai/mlc-llm/pull/506 .

I found this bug is quite interesting. The issue might be in the "FuseDecodeTranspose" pass, de-compressing the decode and transpose block in current way may have some issue.

https://github.com/mlc-ai/mlc-llm/blob/b71bd391239343d262c9931343e20a6aa8ce5bdc/mlc_llm/transform/decode_transpose.py#L84-L94

Which will cause it to trace to the wrong block when applying the default GPU schedule. For example, a transposed-decode func has two blocks, decode & transpose.

```python
@T.prim_func
    def decode2(A: T.Buffer((T.int64(512), T.int64(11008)), "uint32"), B: T.Buffer((T.int64(1), T.int64(1376)), "uint32"), C: T.Buffer((T.int64(1), T.int64(11008)), "float16"), D: T.Buffer((T.int64(4096),), "uint32"), T_transpose: T.Buffer((T.int64(11008), T.int64(4096)), "float16")):
        T.func_attr({"tir.noalias": T.bool(True)})
        # with T.block("root"):
        decode = T.alloc_buffer((T.int64(4096), T.int64(11008)), "float16")
        for i, j in T.grid(T.int64(4096), T.int64(11008)):
            with T.block("decode"):
                v_i, v_j = T.axis.remap("SS", [i, j])
                T.reads(A[v_i // T.int64(8), v_j], C[D[v_i], v_j], D[v_i], B[D[v_i], v_j // T.int64(8)])
                T.writes(decode[v_i, v_j])
                decode[v_i, v_j] = T.Cast("float16", T.bitwise_and(T.shift_right(A[v_i // T.int64(8), v_j], T.Cast("uint32", v_i % T.int64(8) * T.int64(4))), T.uint32(15))) * C[D[v_i], v_j] + T.Cast("float16", T.bitwise_and(T.shift_right(B[D[v_i], v_j // T.int64(8)], T.Cast("uint32", v_j % T.int64(8) * T.int64(4))), T.uint32(15)))
        for ax0, ax1 in T.grid(T.int64(11008), T.int64(4096)):
            with T.block("T_transpose"):
                v_ax0, v_ax1 = T.axis.remap("SS", [ax0, ax1])
                T.reads(decode[v_ax1, v_ax0])
                T.writes(T_transpose[v_ax0, v_ax1])
                T_transpose[v_ax0, v_ax1] = decode[v_ax1, v_ax0]
```

after this pass, we have two funcs, decode&transpose, decode_only.

```python
@T.prim_func
    def decode2(A: T.Buffer((T.int64(512), T.int64(11008)), "uint32"), B: T.Buffer((T.int64(1), T.int64(1376)), "uint32"), C: T.Buffer((T.int64(1), T.int64(11008)), "float16"), D: T.Buffer((T.int64(4096),), "uint32"), T_transpose: T.Buffer((T.int64(11008), T.int64(4096)), "float16")):
        T.func_attr({"tir.noalias": T.bool(True)})
        # with T.block("root"):
        decode = T.alloc_buffer((T.int64(4096), T.int64(11008)), "float16")
        for i, j in T.grid(T.int64(4096), T.int64(11008)):
            with T.block("decode"):
                ...
        for ax0, ax1 in T.grid(T.int64(11008), T.int64(4096)):
            with T.block("T_transpose"):
               ...

@T.prim_func
    def decode6(A: T.Buffer((T.int64(1376), T.int64(4096)), "uint32"), B: T.Buffer((T.int64(1), T.int64(512)), "uint32"), C: T.Buffer((T.int64(1), T.int64(4096)), "float16"), D: T.Buffer((T.int64(11008),), "uint32"), decode: T.Buffer((T.int64(11008), T.int64(4096)), "float16")):
        # with T.block("root"):
        for i, j in T.grid(T.int64(11008), T.int64(4096)):
            with T.block("decode"):
                v_i, v_j = T.axis.remap("SS", [i, j])
                T.reads(A[v_i // T.int64(8), v_j], C[D[v_i], v_j], D[v_i], B[D[v_i], v_j // T.int64(8)])
                T.writes(decode[v_i, v_j])
                decode[v_i, v_j] = T.Cast("float16", T.bitwise_and(T.shift_right(A[v_i // T.int64(8), v_j], T.Cast("uint32", v_i % T.int64(8) * T.int64(4))), T.uint32(15))) * C[D[v_i], v_j] + T.Cast("float16", T.bitwise_and(T.shift_right(B[D[v_i], v_j // T.int64(8)], T.Cast("uint32", v_j % T.int64(8) * T.int64(4))), T.uint32(15)))
```

but when we create schedule, it will track the transpose's parent's to find LCA

https://github.com/apache/tvm/blob/dc7125b31ee456f06d6cf215a40c7ffec83a7532/src/tir/schedule/state.cc#L247-L256 

when it trace the `T_transpose` block's LCA, it will trace to the outside decode_only block(decode6) but the expected one is decode & transpose block (decode2).

One thing we might confuse is why current group quantization doesn't have this issue? That is because the FuseOps and FuseDecodeMatmulEwise pass will legalize the mod, but it seems like these two passes will not apply to other decode functions (like the gptq decode).

To reproduce this issue, you don't indeed need the gptq quantized model, just comment the pass `FuseOps` and `FuseDecodeMatmulEwise`, and build with `python ./build.py --model vicuna-v1-7b --target cuda --quantization q4f16_0 --use-cache=0 --debug-dump`

one possible solution I have made is lift the decode in this way: 

```python
old_body = decode_tir_func.body.block.body[0]
new_body = tir.For(
    loop_var=old_body.loop_var,
    min_val=old_body.min,
    extent=old_body.extent,
    kind=old_body.kind,
    thread_binding=old_body.thread_binding,
    annotations=old_body.annotations,
    body=tir.For(
        loop_var=old_body.body.loop_var,
        min_val=old_body.body.min,
        extent=old_body.body.extent,
        kind=old_body.body.kind,
        thread_binding=old_body.body.thread_binding,
        annotations=old_body.body.annotations,
        body=tir.BlockRealize(
            iter_values=old_body.body.body.iter_values,
            predicate=old_body.body.body.predicate,
            block=tir.Block(
                iter_vars=old_body.body.body.block.iter_vars,
                reads=old_body.body.body.block.reads,
                writes=old_body.body.body.block.writes,
                name_hint=old_body.body.body.block.name_hint,
                body=old_body.body.body.block.body,
            ),
        ),
    ),
)

new_func = tir.PrimFunc(
    params=new_func_buffers,
    body=tir.BlockRealize(
        iter_values=[],
        predicate=True,
        block=tir.Block(
            iter_vars=[],
            reads=[],
            writes=[],
            name_hint="root",
            body=new_body,
        ),
    ),
)
```

This solution is ad hoc as I'm not sure if there is a deep copy like function for TIR stmt. But this indeed works for me.

**Comment by yzh119** [comment]
Currently, TVM runtime objects can be deep copied via `LoadJSON(SaveJSON(obj))`

**Comment by LeiWang1999** [comment]
looks like this issue was fixed through some fuseops related update within unity

**Comment by MasterJH5574** [comment]
> looks like this issue was fixed through some fuseops related update within unity

It must be hidden for some unknown reason before. Now it should be fixed by #535 :-)
